### DISCUSSION TITLE
How to set up development environment for java custom scripts

### DISCUSSION THREAD

**Discussion by polsson12** [JanssenProject/jans#9489] [Q&A] [answered]
Hi,

Can someone provide guidance on how to set up a development environment when writing custom scripts in java? 
When writing custom scripts in java you are using a lot of internal libraries/classes/interfaces etc of janssen, e.g., `SimpleCustomProperty`,  `CustomScript`, `DiscoveryType` etc in the example below. How can I add janssen jars to my class path so it becomes easier and error-free to develop custom scripts locally on my computer?

I've tried to search the web for such jars but can't find any. "Normally" you would for example define dependencies in maven or something similar so that the IDE understands the context of your code.

```
import io.jans.model.SimpleCustomProperty;
import io.jans.model.custom.script.model.CustomScript;

public class Discovery implements DiscoveryType {

    private static final Logger log = LoggerFactory.getLogger(Discovery.class);
    private static final Logger scriptLogger = LoggerFactory.getLogger(CustomScriptManager.class);

    @Override
    public boolean init(Map<String, SimpleCustomProperty> configurationAttributes) {
        log.info("Init of Discovery Java custom script");
        return true;
    }
}
```

**Comment by yuriyz** [comment, marked as answer]
One of the ways it to pick up `jans-auth-server` project with server module in IDE and create script inside it.

https://github.com/JanssenProject/jans/tree/main/jans-auth-server/server

You can see maven pom.xml with all dependencies in it.

**Comment by polsson12** [comment]
If I understand you correctly you suggest to download the jannsen github project and create my scripts inside there to get the context. Is that correct?

Is there no other way to do it? Your suggestion would mean that development is done somewhere, and then files has to be moved around for version control and deployment (we're using terraform) - sounds like a lot of friction and not very intuitive. Additionally we want to have automatic execution of unit tests of our custom scripts in a CI/CD pipeline and I don't see how that can be solved in any easy way with this approach.

**Reply by yuriyz** [reply]
`DiscoveryType` is custom script which is run inside AS. It's not something that lives outside of AS during runtime. Thus the easiest is to develop it in context of AS. 

I see what you are saying about project. It is the easiest to look into AS and look around. But if you are not up to it you can just add maven dependency. Maven repo is here: https://jenkins.jans.io/maven 

E.g. https://jenkins.jans.io/maven/io/jans/jans-auth-server/

Re: CI/CD, once it's developed it will be pushed to DB and picked up by AS.

**Comment by nynymike** [comment]
There is a lot of data injected into the Context which would be hard to replicate in your IDE. If you can come up with a better solution, we're all ears.  Maybe @ossdhaval can point you to some better instructions.
